packages/typescript-fetch-runtime/src/main.ts

```diff
@@ -53,31 +65,87 @@ export type HeaderParams = Record<string, string | undefined>
 export abstract class AbstractFetchClient {
   protected readonly basePath: string
   protected readonly defaultHeaders: Record<string, string>
+  protected readonly defaultTimeout?: number
 
   protected constructor(config: AbstractFetchClientConfig) {
     this.basePath = config.basePath
     this.defaultHeaders = config.defaultHeaders
+    this.defaultTimeout = config.defaultTimeout
+  }
+
+  protected _fetch<R extends Res<any, any>>(
+    url: string,
+    opts: RequestInit,
+    timeout: number | undefined = this.defaultTimeout,
+  ): TypedFetchResponse<R> {
+    // main abort controller that will be returned to the caller
+    const cancelRequest = new AbortController()
+
+    // if a timeout is provided, set a timeout signal and connect to main abort controller
+    if (timeout && timeout > 0) {
+      const timeoutSignal = AbortSignal.timeout(timeout)
+      timeoutSignal.addEventListener("abort", () =>
+        cancelRequest.abort(timeoutSignal.reason),
+      )
+    }
+
+    // if we were provided a signal connect it to the main abort controller
+    const userSignal = opts.signal
+
+    if (userSignal) {
+      userSignal.addEventListener("abort", () =>
+        cancelRequest.abort(userSignal.reason),
+      )
+    }
+
+    const res = fetch(url, {
+      ...opts,
+      signal: cancelRequest.signal,
+    })
+      .then((res) => {
+        cancelRequest.abort()
+        return res
+      })
+      .catch((err) => {
+        // workaround bug where eg: TimeoutError gets converted to an AbortError
+        // https://stackoverflow.com/a/75973817/1445636
+        if (err?.name === "AbortError") {
+          cancelRequest.signal.throwIfAborted()
+        }
+        // if not aborted just throw
+        throw err
+      })
+
+    // decorate the Promise with itself and the AbortController, this allows the
+    // caller to choose between:
+    // - await the Promise and ignore the AbortController
+    // - destructure the Promise to get the Promise and the AbortController
+    // @ts-ignore
```

Comment:
Little bit hacky, but it allows for an API that lets you do either:
```
const res = await this._fetch(..)

if(res.ok) ...
```
Or:
```
const {res, cancelRequest} = this._fetch(...)

btCancel.addEventListener('click', () => cancelRequest.abort())

const response = await res
...
```

Though I'm not loving the name `res` when destructuring, perhaps `whenRes` might be better